electrum/blockchain.py

```diff
@@ -625,9 +646,37 @@ def check_header(header: dict) -> Optional[Blockchain]:
     return None
 
 
-def can_connect(header: dict) -> Optional[Blockchain]:
+def can_connect(header: dict, proof_was_provided: bool=False) -> Optional[Blockchain]:
     with blockchains_lock: chains = list(blockchains.values())
     for b in chains:
-        if b.can_connect(header):
+        if b.can_connect(header, proof_was_provided=proof_was_provided):
             return b
     return None
+
+def verify_proven_chunk(chunk_base_height, chunk_data):
+    chunk = HeaderChunk(chunk_base_height, chunk_data)
+
+    header_count = chunk.get_header_count()
+    prev_header = None
+    prev_header_hash = None
+    for i in range(header_count):
+        header = chunk.get_deserialized_header_at_index(i)
+        # Check the chain of hashes for all headers preceding the proven one.
+        this_header_hash = hash_header(header)
+        if i > 0:
+            if prev_header_hash != header.get('prev_block_hash'):
+                raise Exception("prev hash mismatch: %s vs %s" % (prev_header_hash, header.get('prev_block_hash')))
+        prev_header_hash = this_header_hash
+
+# Copied from electrumx
+def root_from_proof(hash, branch, index):
```

Comment:
note: this looks really similar to https://github.com/spesmilo/electrum/blob/cb204dd969955e6ac6b23821500c406678a3ce97/electrum/verifier.py#L140
but I guess because of `_raise_if_valid_tx`, it might need to be duplicated

Reply:
Hmm... do you think it would be useful to refactor `hash_merkle_root` so that the raise is disableable?

On a related note, do you think it would be useful to validate the length of each member of `branch`?  The length isn't currently validated in either `root_from_proof` or `hash_merkle_root`, and I feel a bit nervous about the potential for attacks possibly being enabled by sending a Merkle branch whose members aren't 32 bytes long.  (For example, what happens if a branch member is the empty string?  Would that result in a Merkle root that's equal to its child?  Can we be confident that this doesn't enable any attacks?)

Reply:
Sorry, I must have missed this message before.
Good point on extra validation; does not seem trivial to come up with an attack but no pointing in risking it.
See https://github.com/spesmilo/electrum/commit/9b82321fc083f3fb4728d9f3785a04861a617011

I think it would be nice not to duplicate this function; so yes, the raise could be made optional based on a parameter.

Reply:
Fixed the duplicated function in latest commit.